ca/certificate-authority.go

```diff
@@ -172,8 +202,40 @@ func loadKey(keyConfig cmd.KeyConfig) (priv crypto.Signer, err error) {
 	return
 }
 
+// checkHSMFault checks whether there has been an HSM fault observed within the
+// timeout window.  CA methods that use the HSM should call this method right
+// away, to minimize the performance impact of HSM outages.
+func (ca *CertificateAuthorityImpl) checkHSMFault() bool {
+	ca.hsmFaultLock.Lock()
+	defer ca.hsmFaultLock.Unlock()
+
+	now := ca.Clk.Now()
+	timeout := ca.hsmFaultLastObserved.Add(ca.hsmFaultTimeout)
+	return now.Before(timeout)
+}
+
+// noteHSMFault updates the CA's state with regard to HSM faults.  CA methods
+// that use an HSM should pass errors that might be HSM errors to this method.
+func (ca *CertificateAuthorityImpl) noteHSMFault(err error) {
+	ca.hsmFaultLock.Lock()
+	defer ca.hsmFaultLock.Unlock()
+
+	if err != nil {
+		ca.stats.Inc(metricHSMFaultObserved, 1, 1.0)
+		ca.hsmFaultLastObserved = ca.Clk.Now()
+	}
+	return
+}
+
 // GenerateOCSP produces a new OCSP response and returns it
 func (ca *CertificateAuthorityImpl) GenerateOCSP(xferObj core.OCSPSigningRequest) ([]byte, error) {
+	if ca.checkHSMFault() {
+		err := core.ServiceUnavailableError("GenerateOCSP call rejected; HSM is unavailable")
+		ca.log.WarningErr(err)
```

Comment:
You can de-dupe the two stanzas that start with `if ca.checkHSMFault()` by making `checkHSMFault` return an error, and take care of the logging and stats stuff internally. The error message would have to be more generic, e.g. "HSM is unavailable," but I think that's fine.